File: ProgrammingWCFServices/Chapter07/PerSessionReleaseInstanceOnComplete/PerSessionReleaseInstanceOnCompleteService/CServiceHostWrapperCollection.cs
using System;
using System.Collections.Generic;
using System.Text;
//-------------------------------------------------------------------------------------------------------
namespace PerSessionReleaseInstanceOnCompleteService
{
//-------------------------------------------------------------------------------------------------------
	public class CServiceHostWrapperCollection
	{
//-------------------------------------------------------------------------------------------------------
		private object											MSyn;
		private List<IServiceHostWrapper>						MList;
//-------------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------------
		public CServiceHostWrapperCollection()
		{
			MSyn=new object();
			MList=new List<IServiceHostWrapper>();
		}
//-------------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------------
		public void StartServices()
		{
			lock(MSyn)
			{
				foreach(IServiceHostWrapper ServiceHostWrapper in MList)
					ServiceHostWrapper.StartService();
			}
		}
//-------------------------------------------------------------------------------------------------------
		public void StopServices()
		{
			lock(MSyn)
			{
				foreach(IServiceHostWrapper ServiceHostWrapper in MList)
					ServiceHostWrapper.StopService();
			}
		}
//-------------------------------------------------------------------------------------------------------
		public void AddServiceHostWrapper(IServiceHostWrapper ServiceHostWrapper)
		{
			lock(MSyn)
			{
				MList.Add(ServiceHostWrapper);
			}
		}
//-------------------------------------------------------------------------------------------------------
		public bool RemoveServiceHostWrapper(int Index)
		{
			lock(MSyn)
			{
				if (Index>=MList.Count)
					return(false);

				MList.RemoveAt(Index);
				return(true);
			}
		}
//-------------------------------------------------------------------------------------------------------
	}
//-------------------------------------------------------------------------------------------------------
}
//-------------------------------------------------------------------------------------------------------

File: ProgrammingWCFServices/CPP/SimpleWCF/SimpleWCFService/CServiceHostWrapperCollection.cpp
#include "StdAfx.h"
#include "CServiceHostWrapperCollection.h"
//-------------------------------------------------------------------------------------------------------
namespace SimpleWCFService
{
//-------------------------------------------------------------------------------------------------------
CServiceHostWrapperCollection::CServiceHostWrapperCollection(void)
{
	MSyn=gcnew Object();
	MList=gcnew List<IServiceHostWrapper^>;
}
//-------------------------------------------------------------------------------------------------------
CServiceHostWrapperCollection::~CServiceHostWrapperCollection(void)
{
	StopServices();
}
//-------------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------------
void CServiceHostWrapperCollection::StartServices(void)
{
	lock			MLockRoot(this);
	{
		for each(IServiceHostWrapper^ ServiceHostWrapper in MList)
			ServiceHostWrapper->StartService();
	}
}
//-------------------------------------------------------------------------------------------------------
void CServiceHostWrapperCollection::StopServices(void)
{
	lock			MLockRoot(this);
	{
		for each(IServiceHostWrapper^ ServiceHostWrapper in MList)
			ServiceHostWrapper->StopService();
	}
}
//-------------------------------------------------------------------------------------------------------
void CServiceHostWrapperCollection::AddServiceHostWrapper(IServiceHostWrapper^ ServiceHostWrapper)
{
	lock			MLockRoot(this);
	{
		MList->Add(ServiceHostWrapper);
	}
}
//-------------------------------------------------------------------------------------------------------
bool CServiceHostWrapperCollection::RemoveServiceHostWrapper(int Index)
{
	lock			MLockRoot(this);
	{
		if (Index>=MList->Count)
			return(false);

		MList->RemoveAt(Index);
		return(true);
	}
}
//-------------------------------------------------------------------------------------------------------
}
//-------------------------------------------------------------------------------------------------------